File: 0101-symmetric-tree/0101-symmetric-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool temp(TreeNode* left, TreeNode* right){
        if(!left && !right) return true;
        if(!left || !right) return false;
        if(left->val != right->val) return false;
        return temp(left->left, right->right) && temp(left->right, right->left);
    }

    bool isSymmetric(TreeNode* root) {
        if(!root) return false;
        return temp(root->left, root->right);
    }
};

File: 0101-symmetric-tree/0101-symmetric-tree.dart
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *   int val;
 *   TreeNode? left;
 *   TreeNode? right;
 *   TreeNode([this.val = 0, this.left, this.right]);
 * }
 */
class Solution {
  bool isSymmetric(TreeNode? root) {
    if(root == null) return true;
    else if(root.left == null && root.right == null) return true;
    else if(root.left == null || root.right == null) return false; 
    return solve(root.left, root.right);
  }
  bool solve(TreeNode? left, TreeNode? right){
    if(left == null && right == null) return true;
    else if (left == null || right == null) return false;
    else if(left.val != right.val) return false;
    return solve(left.left, right.right) && solve(right.left, left.right);
  }
}